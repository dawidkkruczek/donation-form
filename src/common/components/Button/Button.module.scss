.root {
  background-color: transparent;
  border-radius: 5px;
  border: 1px solid transparent;
  font-family: 'Work Sans';
  font-size: 16px;
  font-weight: 600;
  justify-content: center;
  line-height: 1.25;
  padding: 16px 24px;
  transition: all 0.5s ease;

  &:disabled {
    opacity: 0.75;
  }

  &:not(:disabled) {
    cursor: pointer;
  }
}

.rootPrimary {
  background-color: var(--color-midnight-purple);
  color: var(--color-white);

  &:not(:disabled) {
    &:hover {
      background-color: var(--color-midnight-purple-light);
    }

    &:active {
      background-color: var(--color-midnight-purple-dark);
    }
  }
}

.rootGhost {
  border-color: var(--color-purple-gray);
  color: var(--color-purple-gray);

  &:not(:disabled) {
    &:hover {
      background-color: var(--color-ghost-purple-light);
    }

    &:active {
      background-color: var(--color-ghost-purple-dark);
    }
  }
}

.rootFullWidth {
  width: 100%;
}
