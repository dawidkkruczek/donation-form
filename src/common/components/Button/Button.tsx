import React, { ComponentProps, ReactElement } from 'react';
import { pascalCase } from 'change-case';
import classNames from 'classnames';
import styles from './Button.module.scss';

enum Variant {
  Ghost = 'ghost',
  Primary = 'primary',
}
export interface Props extends ComponentProps<'button'> {
  fullWidth?: boolean;
  onClick?: () => void;
  variant?: Variant;
}

function Button({
  children,
  className,
  fullWidth,
  type = 'button',
  variant = Variant.Primary,
  ...buttonProps
}: Props): ReactElement {
  const rootClassName = classNames(styles.root, className, {
    [styles.rootFullWidth]: fullWidth,
    [styles[`root${pascalCase(variant)}`]]: variant,
  });

  return (
    <button className={rootClassName} type={type} {...buttonProps}>
      {children}
    </button>
  );
}

Button.Variant = Variant;

export default Button;
