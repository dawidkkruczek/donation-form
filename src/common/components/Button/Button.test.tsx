import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Button from './Button';

describe('common/components/Button', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<Button>Sample</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with given style', () => {
    Object.values(Button.Variant).forEach((variant) => {
      const { asFragment } = render(<Button variant={variant}>Sample</Button>);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  it('should render correctly with fullWidth', () => {
    const { asFragment } = render(<Button fullWidth>Sample</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should trigger onClick when passed', () => {
    const onClick = jest.fn();
    const { getByText } = render(<Button onClick={onClick}>Sample</Button>);
    fireEvent.click(getByText('Sample'));
    expect(onClick).toHaveBeenCalled();
  });
});
