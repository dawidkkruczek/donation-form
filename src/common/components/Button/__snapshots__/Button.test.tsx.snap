// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/components/Button should render correctly 1`] = `
<DocumentFragment>
  <button
    class="root rootPrimary"
    type="button"
  >
    Sample
  </button>
</DocumentFragment>
`;

exports[`common/components/Button should render correctly with fullWidth 1`] = `
<DocumentFragment>
  <button
    class="root rootFullWidth rootPrimary"
    type="button"
  >
    Sample
  </button>
</DocumentFragment>
`;

exports[`common/components/Button should render correctly with given style 1`] = `
<DocumentFragment>
  <button
    class="root rootGhost"
    type="button"
  >
    Sample
  </button>
</DocumentFragment>
`;

exports[`common/components/Button should render correctly with given style 2`] = `
<DocumentFragment>
  <button
    class="root rootPrimary"
    type="button"
  >
    Sample
  </button>
</DocumentFragment>
`;
